xarray/test/test_backends.py

```diff
@@ -912,7 +913,12 @@ def test_cross_engine_read_write_netcdf3(self):
                     for read_engine in valid_engines:
                         with open_dataset(tmp_file,
                                           engine=read_engine) as actual:
-                            self.assertDatasetAllClose(data, actual)
+                            # hack to allow test to work:
```

Comment:
This is fine for now, but we should look into this later because it looks like a legit bug to me.

It might be worth making a dedicated `VariablesDict` that is simply an OrderedDict with runtime type checking that verifies you can never put in anything other than a `Variable`. Or we should use pytype for static checks to ensure `Dataset._variables` and `DataArray._coords` are typed `typing.OrderedDict[Any, xarray.Variable]`.